java/arcs/sdk/HandleUtils.kt

```diff
@@ -31,19 +31,353 @@ suspend fun <T1, T2> combineUpdates(
+    )
+    handles.forEach { handle ->
+        handle.onUpdate {
+            val e1 = handle1.getContent()
+            val e2 = handle2.getContent()
+            val e3 = handle3.getContent()
+            val e4 = handle4.getContent()
+            val e5 = handle5.getContent()
+            val e6 = handle6.getContent()
+            val e7 = handle7.getContent()
+            val e8 = handle8.getContent()
+            action(e1, e2, e3, e4, e5, e6, e7, e8)
+        }
+    }
+}
+
+/**
+ * Receive a callback when either handle is updated.
+ *
+ * @handle1 The first handle the callback will be assigned to
+ * @handle2 The second handle the callback will be assigned to
+ * @handle3 The third handle the callback will be assigned to
+ * @handle4 The fourth handle the callback will be assigned to
+ * @handle5 The fifth handle the callback will be assigned to
+ * @handle6 The sixth handle the callback will be assigned to
+ * @handle7 The seventh handle the callback will be assigned to
+ * @handle8 The eighth handle the callback will be assigned to
+ * @handle9 The ninth handle the callback will be assigned to
+ * @action callback
+ */
+suspend fun <T1, T2, T3, T4, T5, T6, T7, T8, T9> combineUpdates(
+    handle1: ReadableHandle<T1, *>,
+    handle2: ReadableHandle<T2, *>,
+    handle3: ReadableHandle<T3, *>,
+    handle4: ReadableHandle<T4, *>,
+    handle5: ReadableHandle<T5, *>,
+    handle6: ReadableHandle<T6, *>,
+    handle7: ReadableHandle<T7, *>,
+    handle8: ReadableHandle<T8, *>,
+    handle9: ReadableHandle<T9, *>,
+    action: (T1, T2, T3, T4, T5, T6, T7, T8, T9) -> Unit
+) {
+    val handles = listOf(
+        handle1,
+        handle2,
+        handle3,
+        handle4,
+        handle5,
+        handle6,
+        handle7,
+        handle8,
+        handle9
+    )
+    handles.forEach { handle ->
+        handle.onUpdate {
+            val e1 = handle1.getContent()
+            val e2 = handle2.getContent()
+            val e3 = handle3.getContent()
+            val e4 = handle4.getContent()
+            val e5 = handle5.getContent()
+            val e6 = handle6.getContent()
+            val e7 = handle7.getContent()
+            val e8 = handle8.getContent()
+            val e9 = handle9.getContent()
+            action(e1, e2, e3, e4, e5, e6, e7, e8, e9)
+        }
+    }
+}
+
+/**
+ * Receive a callback when either handle is updated.
+ *
+ * @handle1 The first handle the callback will be assigned to
+ * @handle2 The second handle the callback will be assigned to
+ * @handle3 The third handle the callback will be assigned to
+ * @handle4 The fourth handle the callback will be assigned to
+ * @handle5 The fifth handle the callback will be assigned to
+ * @handle6 The sixth handle the callback will be assigned to
+ * @handle7 The seventh handle the callback will be assigned to
+ * @handle8 The eighth handle the callback will be assigned to
+ * @handle9 The ninth handle the callback will be assigned to
+ * @action callback
```

Comment:
I think these are probably not necessary.

...But if you think they are, it should be `@param handle1 blah blah`

Reply:
+1. I'd suggest keeping the param descriptions for the first `combineUpdates` impl, then omitting them thereafter. It's pretty clear what the interface is.

If there's a lint requirement to have doc comments, just put `/** N-handle version of combineUpdates */` for N > 2.

Reply:
> +1 to Piotr's suggestion.
> 
> Possible alternative for avoiding so much repetition: could we have convenience impls for 2-4 handles, and then one that just accepts a list of handles? I mean, you're already building that list internally, so is it that hard for callers to say `combineUpdates(listOf(...))` for N > 4?

This was one of the alternatives we considered (as was varArg), I even did a prototype implementation. We went away from it, because we wanted to provide the updated (and typed) entities or sets of entities (based on whether it was a Singleton or Collection handle) to the callback.

Reply:
FWIW, I do think having properly typed contents of handles available is a huge improvements over just registering callbacks and having the developer fetch the results themselves.

Reply:
Yes, pre-fetching with correct types on the output side is definitely worth the repetition here. :+1: